cpp/src/io/csv/csv_gpu.cu

```diff
@@ -570,6 +571,329 @@ __global__ void convertCsvToGdf(const char *raw_csv,
+  c_prev                            = (cur > start) ? cur[-1] : terminator;
+  // Loop through all 32 bytes and keep a bitmask of row starts for each possible input context
+  for (uint32_t pos = 0; pos < 32; pos++, cur++, c_prev = c) {
+    uint32_t ctx;
+    if (cur < end) {
+      c = cur[0];
+      if (c_prev == terminator) {
+        if (c == commentchar ||
+            (skipblanklines && (c == terminator || (c == '\r' && terminator == '\n')))) {
+          // Start of a new comment or empty row
+          ctx = make_short_row_context(ROW_CTX_COMMENT, ROW_CTX_QUOTE, ROW_CTX_COMMENT, 0, 0, 0);
+        } else if (c == quotechar) {
+          // Quoted string on newrow, or quoted string ending in terminator
+          ctx = make_short_row_context(ROW_CTX_QUOTE, ROW_CTX_NONE, ROW_CTX_QUOTE, 1, 0, 1);
+        } else {
+          // Start of a new row unless within a quote
+          ctx = make_short_row_context(ROW_CTX_NONE, ROW_CTX_QUOTE, ROW_CTX_NONE, 1, 0, 1);
+        }
+      } else if (c == quotechar) {
+        if (c_prev == delimiter || c_prev == quotechar) {
+          // Quoted string after delimiter, quoted string ending in delimiter, or double-quote
+          ctx = make_short_row_context(ROW_CTX_QUOTE, ROW_CTX_NONE);
+        } else {
+          // Closing or ignored quote
+          ctx = make_short_row_context(ROW_CTX_NONE, ROW_CTX_NONE);
+        }
+      } else if (c == delimiter_esc && c_prev != escapechar) {
+        // Escapable delimiter ends quoted string
+        ctx = make_short_row_context(ROW_CTX_NONE, ROW_CTX_NONE);
+      } else {
+        // Neutral character
+        ctx = make_short_row_context(ROW_CTX_NONE, ROW_CTX_QUOTE);
+      }
+    } else {
+      const char *data_end = start + data_size - start_offset;
+      if (cur >= data_end) {
+        // Add a newline at data end (need the extra row offset to infer length of previous row)
+        uint32_t eof_row = (cur == data_end && cur <= end) ? 1 : 0;
+        ctx =
+          make_short_row_context(ROW_CTX_EOF, ROW_CTX_EOF, ROW_CTX_EOF, eof_row, eof_row, eof_row);
+      } else {
+        // Pass-through context (beyond chunk_size but before data_size)
+        ctx = make_short_row_context(ROW_CTX_NONE, ROW_CTX_QUOTE, ROW_CTX_COMMENT);
+      }
+    }
+    // Merge with current context, keeping track of where new rows occur
+    merge_short_row_context(ctx_map, ctx, pos);
+  }
+
+  // Eliminate rows that start before byte_range_start
+  if (start_offset + block_pos < byte_range_start) {
+    uint32_t dist_minus1 = min(byte_range_start - (start_offset + block_pos) - 1, UINT64_C(31));
+    uint32_t mask        = 0xfffffffe << dist_minus1;
+    ctx_map.x &= mask;
+    ctx_map.y &= mask;
+    ctx_map.z &= mask;
+  }
+
+  // Convert the long-form {rowmap,outctx}[inctx] version into packed version
+  // {rowcount,ouctx}[inctx]
+  ctxb = pack_row_contexts(make_row_context(__popc(ctx_map.x), (ctx_map.w >> 0) & 3),
```

Comment:
This could be an inlined utility function, `pack_short_row_context` or something

Reply:
```cpp
ctxb = pack_row_contexts(make_row_context(__popc(ctx_map.x), (ctx_map.w >> 0) & 3),
                           make_row_context(__popc(ctx_map.y), (ctx_map.w >> 2) & 3),
                           make_row_context(__popc(ctx_map.z), (ctx_map.w >> 4) & 3));
```

Reply:
I just removed `ctxb` in the latest commit, and just called passed the result of `pack_row_context()` directly to `rowctx_merge_transform`, since it was used only for that purpose and only in one place.

Reply:
Would still prefer to see a `pack_short_row_context` and `select_short_row_context` (or maybe `select_rowmap` ?

Reply:
I think it would be good to encapsulate all the low level bit twiddling stuff into short functions, yes. `select_rowmap` sounds good. I think the 32-bit contexts, packed contexts, short context, etc could be bundled up into classes without adding any overhead, but it could be done in a follow up PR (I wouldn't mind playing around with it myself)

Is this the only place where `rowctx_merge_transform` is called? I suppose you could pass `ctx_map` here instead and do the packing inside `rowctx_merge_transform`.

Reply:
Yes, there is only one place where it is called, but the merge transform is a fairly generic operation whereas the short row context stuff is definitely specific to the calling kernel, so I'd rather keep them separate -> I'd rather go with the short util function in this case.

Reply:
Went with `pack_rowmaps()` and `select_rowmap()` to avoid more "row context" confusion. Only concern is if the compiler would infer that ctx_map.w is still needed because the whole uint4 is being passed in to a function (+1 register), but not concerned enough to spend more time on this.